amul_topicals.py

```diff
@@ -49,15 +72,71 @@ def fetch_url(url):
     r = requests.get(url)
     return r.content
 
+def get_UserYesOrNo(message, options):
+    """User Interaction: Input string message and list of options, returns user choice"""
+    choice = raw_input(message + " ("+"/".join(map(str,options))+") :").lower()
+    while choice.lower() not in options:
+        choice=get_UserYesOrNo(message, options)
+    return choice
+
+def MakeFolderInDirectory(location,foldername):
+    """Make new folder called <foldername> in the given location; returns the directory of folder"""
+    location=location.rstrip('/') #removing any slashes at the end if user has put it
+    location=location.rstrip("\\")
+    if not os.path.exists(location+"/"+foldername):
+        print "A folder called " + "'" +foldername+"'"+" has been created in "+location
+        os.makedirs(location+"/"+foldername) #slash is provided here
+    else:
+        print "Folder "+ "'"+foldername+"'"+" is already present in "+location+"; proceeding to next step..."
+    print
+    location = location + "/" + foldername
+    return location #since modification above; also because folder is to be accessed in the next step
+
+def OpenFolder(location):
+    """Opens the folder at the location; works only for Windows, Mac, Linux"""
+    platforms=['win32','linux2','darwin']
+    current_platform=sys.platform
+    if current_platform not in platforms:
+        return
+    query = get_UserYesOrNo("Do you wish to view the folder?", ['y','n'])
+    if query=='y':        
+        #linux
+        if current_platform=='linux2':
+            os.system('xdg-open "%s"' % location)
+        #mac
+        elif current_platform=='darwin':    
+            os.system('open "%s"' % location)
+        #windows
+        elif current_platform=='win32':
+            os.startfile(location)
+        return
+ 
+
 if __name__ == "__main__":
     page = fetch_url(BASE_URL)
     year_urls = get_year_urls(page)
     # print year_urls
     print
     d = {}
+
+    #asking user to enter location to save files
+    choice = get_UserYesOrNo("Save images in current directory? (a new folder will be created in current directory)", ['y','n'])    
+    if choice.lower() == 'y':
+        location = CWD #storing images in current working directory (default)
+    elif choice.lower() == 'n':
+        #asking user for entering directory
+        location = raw_input("Enter a directory to save images (a new folder will be created in the directory mentioned):")
+        while not os.path.isdir(location):
+            location = raw_input("Not a valid directory, please enter again:")
+    #make new folder called 'Amul Topicals' in the directory; files will be stored in this folder
+    location = MakeFolderInDirectory(location, 'Amul Topicals')    
+    
     for name, url in year_urls.iteritems():
```

Comment:
Here it will be a good idea to sort the years in advance so the user is not presented with a random order of years, e.g. see below on the years I was asked:
```
Do you wish to download images of - Amul hits of 1976? (y/n) :y
...
Do you wish to download images of - Amul hits of 1986? (y/n) :y
...
Do you wish to download images of - Amul hits of 1987? (y/n) :y
...
Do you wish to download images of - Amul hits of 1982? (y/n) :n

Amul hits of 1982 page 1 successfully accessed.


Amul hits of 1982 page 2 successfully accessed.


Amul hits of 1982 page 3 successfully accessed.

Do you wish to download images of - Amul hits of 1983? (y/n) :n

Amul hits of 1983 page 1 successfully accessed.


Amul hits of 1983 page 2 successfully accessed.


Amul hits of 1983 page 3 successfully accessed.


Amul hits of 1983 page 4 successfully accessed.

Do you wish to download images of - Amul hits of 1981? (y/n) :^C
```

Further, the output lines such `Amul hits of 1983 page 3 successfully accessed` for years where user selected no are misleading.

Reply:
This was something I added in my previous commit, where we had to extend the program to fetch all the topicals. I did mention in a note that I didn't destroy this feature of fetching all the topicals, since at the end it stores everything into 'json_data'. Should I get rid of it, or store it anyway without showing any output lines like you've mentioned? If we have to still keep that feature, one thing to be noted is that if the program will run for that part without printing the 'successfully accessed' lines or any such output for that matter, that time will still be taken and it would seem like a 'pause' to the user. Please suggest a way to go by this.

Reply:
Well, what do you think will a user want?

I can think of two possibilities:
1. Fetching all topicals (with or without progress shown is your choice).
2. Just the one year they ask for.

Reply:
Understood. The user wouldn't really want what they don't ask for, so the latter one it is.